core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -724,6 +756,23 @@ public void kill() throws InterruptedException {
                 int pid = getPid();
                 LOGGER.fine("Killing pid="+pid);
                 UnixReflection.destroy(pid);
+                // after sending SIGTERM, wait for the process to cease to exist
+                long deadline = System.nanoTime() + softkillWaitSeconds * 1000000000;
+                do {
+                    try {
+                        FileReader r = new FileReader(getFile("status"));
+                        r.close();
+                    }
+                    catch (Exception e) {
+                        break; // process probably no longer exists (status gone)
+                    }
+
+                    try {
+                        Thread.sleep(1000);
```

Comment:
It could check more often for first few seconds. Otherwise it may cause a performance impact when aborting process trees

Reply:
Good idea! I'll make it sleep 10, 20, 40, 80 ... up to 1000 msecs.